src/lib/lt.js

```diff
@@ -0,0 +1,64 @@
+const crypto = require('crypto')
+const IlpPacket = require('ilp-packet')
+const IlDcp = require('ilp-protocol-ildcp')
+
+function sha256 (preimage) {
+  return crypto.createHash('sha256').update(preimage).digest()
+}
+
+class Loop {
+  constructor (plugin1, plugin2, destination) {
+    this.pending = {}
+    this.destination = destination
+    this.plugin1 = plugin1
+    this.plugin2 = plugin2
+    this.plugin2.registerDataHandler(this._loopbackPrepareHandler.bind(this))
+  }
+
+  async _loopbackPrepareHandler (packet) {
+    const { amount, executionCondition } = IlpPacket.deserializeIlpPrepare(packet)
+    if (this.pending[executionCondition]) {
+      const shouldFulfill = await this.pending[executionCondition].loopbackHandler(amount)
+      if (shouldFulfill) {
+        const fulfillment = this.pending[executionCondition].fulfillment
+        return IlpPacket.serializeIlpFulfill({ fulfillment, data: Buffer.from([]) })
+      }
+    }
+    return IlpPacket.serializeIlpReject({
+      code: 'F04',
+      triggeredBy: this.destination,
+      message: 'Insufficient destination amount',
+      data: Buffer.from([])
+    })
+  }
+
+  async pay ({ sourceAmount, expiresAt, loopbackHandler }) {
+    const fulfillment = crypto.randomBytes(32)
+    const executionCondition = sha256(fulfillment)
+    const packet = IlpPacket.serializeIlpPrepare({
+      amount: sourceAmount,
+      expiresAt,
+      executionCondition,
+      destination: this.destination,
+      data: Buffer.from([])
```

Comment:
There may still be a reason to include data here. What if the sender wants to correlate packets between the sending and receiving sides by putting some kind of ID in the data?

Reply:
Yeah, I had that at first, but then realized that the condition is already a unique identifier (and is also used actually used as such in the code, as the index to the `pending` hashmap).